### DISCUSSION TITLE
Fulltext Search in ETL

### DISCUSSION THREAD

**Discussion by tylkomat** [ravendb/ravendb#15048] [Q&A] [answered]
I'm currently learning RavenDB and use it for data processing. I want to do as much as possible inside of RavenDB (studio) without external scripts (to see the capabilities). 

I have data from a CSV which is not really usable so I was transforming it using an ETL.
Would this also work with an index? I didn't really manage. I was always getting a JINT Exception with statements executed have been reached. It also didn't matter to increase it even to 1 million.

Any way, there is some free text in my data, like (partial) names of persons. 
Is it possible to do a full text search in the ETL to find the document ids of the names of the persons?
If not, are there other options to transform the data into a structure with document ids?

**Comment by ayende** [comment]
I don't understand your query.
You have CSV data as a file? 
Did you import that to RavenDB?
AS a single document, or multiple?

What are you trying to do in Jint?

**Comment by tylkomat** [comment, marked as answer]
Yes, I imported the CSV into RavenDB. It created multiple documents (14).
At first I tried to create an index to map the data into another format. I used Javascript, not RQL. But the index couldn't be created because of the `StatementsCountOverflowException`, while the code wasn't really complex.

I then found out that indexes are not used for transforming data, so I switched to RavenDB ETL. The code was working fine there.

**Comment by arekpalinski** [comment]
I think you were looking for the patching feature:
https://ravendb.net/docs/article-page/5.4/csharp/client-api/operations/patching/set-based
https://ravendb.net/docs/article-page/5.4/csharp/studio/database/documents/patch-view

So you could transform your data locally (in the same db) without the need to ETL them to another db.

**Comment by tylkomat** [comment]
Thanks for the info, but I wanted to keep the original data.

Coming back to my question.
Is it possible to do a full text search in the ETL to find the document ids of the names of persons?
If not, are there other options to transform the data into a structure with document ids?

**Reply by arekpalinski** [reply]
Note you can use `put()` method in the patch script in order to put a new document and keep the original one not modified. For example: `put("newUsers/", { Name: this.FirstName + ' ' + this.LastName)`).

Note that I used `"newUsers/"` as the document ID so it will be generated by server:
https://ravendb.net/docs/article-page/5.4/csharp/client-api/document-identifiers/working-with-document-identifiers#server-side-generated-ids
but you can specify it explicitly.

There isn't an option to do full text search during the JS script execution (neither during a patch script nor ETL transformation script). 

Can you elaborate on your last question? I'm afraid I don't follow it.

**Reply by tylkomat** [reply]
Thank you for the hint.

I have free text values in my json. The free text also mentions names of persons. Sometimes only first name or with abbreviated last name (i.e. John S.) or only last names (i.e. Mr. Smith). 

The data comes from a badly designed questionaire. The question was: Who do you have to cooperate with? An answer would be: I have to ask John to get information about X and deliver my work to Mrs. Jackson.

I now want to extract the names and find corresponding ids, such that I can link to the mentioned people in the database, to augment the data with:
```json
{
  ...
  "cooperatesWith": [
    "people/A-100", // John Smith
    "people/A-101", // Mary Jackson
  ],
  ...
}
```

Full text search would be great to use, but since that is not an option, I was wondering if there is another way.
I'm currently working on a subscription to do that transformation. There I can use full text search to combine the results.

Since I'm just learning RavenDB I wonder if there is another way.

**Reply by arekpalinski** [reply]
That sounds like a problem which you can solve with the usage of RavenDB features but you shouldn't try hard to accomplish it only using RavenDB server. I think you might need to provide some external knowledge to have good matching of mentioned people. 

You might transform your data using external program but using features like full text search querying or patching via RavenDB client.

**Reply by tylkomat** [reply]
Thanks, yes that's what I'm doing now.
